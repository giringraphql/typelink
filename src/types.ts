export type ResolvedValue<T> = T | Promise<T>;
export type ResolvedList<T> = T[] | Promise<T[]> | Promise<T>[];

export interface ConcreteClass<T = any> {
  new(...args: any[]): T;
}

export type Instantiator<TClass = any> = (value: { [key: string]: any }) => TClass;

export function defaultInputFieldInstantiator(value: any) {
  return value;
}

export type Thunk<T> = () => T;
