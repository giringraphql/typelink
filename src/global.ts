import { loadFallbackRootTypes } from './definition/rootTypes';
import { loadBuiltInScalar } from './definition/scalar';
import { DefinitionKind } from './metadata/Definition';
import { MetadataStorage } from './metadata/MetadataStorage';
import { DefinitionParser } from './sdl/ast';
import { coerceType } from './type-expression/coerceType';
import { TypeExpression } from './type-expression/TypeExpression';
import { Thunk } from './types';
import { TypeArg } from './type-expression/types';


/**
 * Create a new MetadataStorage initialized with default metadata
 */
export function createMetadataStorage() {
  const storage = new MetadataStorage();
  loadBuiltInScalar(storage);
  loadFallbackRootTypes(storage);
  return storage;
}

/**
 * Global MetadataStorage used by default.
 */
export let globalMetadataStorage: MetadataStorage;

export function getGlobalMetadataStorage() {
  if (!globalMetadataStorage) {
    globalMetadataStorage = createMetadataStorage();
  }
  return globalMetadataStorage;
}

/**
 * Get a GraphQLType instance from the given storage or default
 * global metadata storage.
 * @param typeArg
 * @param storage
 */
export function getType(typeArg: TypeExpression | TypeArg, kind: DefinitionKind = 'any'): any {
  const storage = getGlobalMetadataStorage();
  return coerceType(typeArg).getType({ kind, storage });
}

/**
 * Define a type linked to decorated class and add it to the given
 * storage or default global metadata storage.
 * @param metadataOrThunk
 */
export function defineType(parsersOrThunk: DefinitionParser[] | Thunk<DefinitionParser[]>) {
  const storage = getGlobalMetadataStorage();
  return function defDecoratorFn<T extends Function>(definitionClass: T) {
    if (Array.isArray(parsersOrThunk)) {
      parsersOrThunk.forEach(parser => {
        storage.registerMetadata(parser.parse(definitionClass));
      });
    } else {
      storage.deferRegister(() => {
        parsersOrThunk().forEach(parser => {
          storage.registerMetadata(parser.parse(definitionClass));
        });
      });
    }
    return definitionClass;
  };
}
