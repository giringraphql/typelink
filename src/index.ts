export { gql } from './sdl/gql';

export * from './definition/GraphQLTypeIndex';
export * from './definition/InputType';
export * from './definition/InterfaceType';
export * from './definition/ObjectType';
export * from './definition/rootTypes';
export * from './definition/scalar';
export * from './definition/SubscriptionType';


export * from './reference/Field';
export * from './reference/Implement';
export * from './reference/InputField';
export * from './global';

export * from './type-expression/DefinitionTypeExpression';
export * from './type-expression/GraphQLTypeExpression';
export * from './type-expression/TypeExpression';
export * from './type-expression/structure';
export * from './type-expression/coerceType';
export * from './type-expression/types';

export * from './metadata/Definition';
export * from './metadata/MetadataStorage';
export * from './metadata/Reference';

export * from './relay/connection';
export * from './relay/mutation';
export * from './relay/node';
