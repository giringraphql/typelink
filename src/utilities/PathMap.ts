export class PathMap<K, V> {
  protected children: Map<K, PathMap<K, V>> = new Map();
  protected value: V;

  get(path: K[]) {
    let submap: PathMap<K, V> = this;
    for (let i = 0; i < path.length; i++) {
      const item = submap.children.get(path[i]);
      if (item === undefined) {
        return undefined;
      } else {
        submap = item;
      }
    }
    return submap.value;
  }
  set(path: K[], value: V) {
    let submap: PathMap<K, V> = this;
    for (let i = 0; i < path.length; i++) {
      let item = submap.children.get(path[i]);
      if (item === undefined) {
        item = new PathMap();
        submap.children.set(path[i], item);
        submap = item;
      } else {
        submap = item;
      }
    }
    submap.value = value;
  }
}
