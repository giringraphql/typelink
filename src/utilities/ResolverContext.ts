import { GraphQLResolveInfo } from 'graphql';


/**
 * Class-based resolver context
 */
export class ResolverContext<TSource = undefined, TContext = any> {
  public $env: IArguments;
  public get $source(): TSource {
    return this.$env[0];
  }
  public get $context(): TContext {
    return this.$env[1];
  }
  public get $info(): GraphQLResolveInfo {
    return this.$env[2];
  }

  constructor(_base: TSource, _context?: TContext, _info?: GraphQLResolveInfo) {
    this.$env = arguments;
  }
}

/**
 * Decorator for fields resolved from source directly
 * @param fieldName
 */
export function source(fieldName?: string) {
  return function(prototype: { $source: any }, propertyKey: string) {
    const get = function(this: any) {
      return this.$source[fieldName || propertyKey];
    };
    const set = function(this: any, value: any) {
      this.$source[fieldName || propertyKey] = value;
    };
    Object.defineProperty(prototype, propertyKey, { get, set });
  };
}

/**
 * Decorator for field resolved after async fetching
 * prototype should have `$fetch()` method which returns a Promise
 * @param fieldName
 */
export function lazy(fieldName?: string) {
  return function(prototype: { $fetch: () => Promise<any> }, propertyKey: string) {
    const get = function(this: any) {
      if (!this.$__fetcher__) {
        this.$__fetcher__ = this.$fetch();
      }
      return this.$__fetcher__.then((remoteSource: any) => remoteSource[fieldName || propertyKey]);
    };
    Object.defineProperty(prototype, propertyKey, { get });
  };
}
