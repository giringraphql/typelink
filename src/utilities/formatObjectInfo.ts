export function formatObjectInfo(obj: any): string {
  let name: string;
  let type: string;
  try {
    if (typeof obj === 'function') {
      name = obj.name || '[anonymous function]';
      type = typeof obj;
    } else {
      name = String(obj);
      type = obj && obj.constructor.name;
    }
    return `${name}<${type}>`;
  } catch (e) {
    return String(obj);
  }
}
