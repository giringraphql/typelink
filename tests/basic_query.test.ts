import { graphql, GraphQLSchema, printSchema } from 'graphql';

import { getType, gql, defineType } from '../src';
import { ResolverContext, source } from '../src/utilities/ResolverContext';


interface TestSource {
  fieldWithDefaultResolver?: string;
}

@defineType(gql`
  type Test {
    """description1"""
    resolverGotDefinitionInstance: Boolean!

    greeting(greeting: String, name: String!): String!
    bigGreeting(greeting: String, name: String!): String!
    fieldWithDefaultResolver: String
  }
`)
class Test extends ResolverContext<TestSource> {
  public resolverGotDefinitionInstance() {
    return this instanceof Test;
  }
  public greeting({ name, greeting }: { name: string, greeting: string }) {
    return `${greeting || 'Hello'}, ${name}`;
  }
  public bigGreeting(arg: { name: string, greeting: string }) {
    return this.greeting(arg) + '!';
  }

  @source() fieldWithDefaultResolver?: string;
}

@defineType(gql`
  type Query {
    test: ${Test}
    erroneousTest: ${Test}
    testPassingSource: ${Test}
  }
`)
class Query extends ResolverContext {
  public static test() {
    return new Test({});
  }
  public static testPassingSource() {
    return new Test({ fieldWithDefaultResolver: 'Ohayo' });
  }
}

const schema = new GraphQLSchema({ query: getType(Query) });


describe('Basic queries and schema generation', async () => {

  it('generates schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('passes source and args to its resolver', async () => {
    let result = await graphql({ schema, source: `
      query {
        test {
          greeting(name: "Suzuki")
          bigGreeting(name: "Suzuki")
          fieldWithDefaultResolver
        }
        testPassingSource {
          fieldWithDefaultResolver
        }
      }
    `});
    expect(result).toEqual({ data: {
      test: {
        greeting: 'Hello, Suzuki',
        bigGreeting: 'Hello, Suzuki!',
        fieldWithDefaultResolver: null,
      },
      testPassingSource: {
        fieldWithDefaultResolver: 'Ohayo',
      },
    } });

    result = await graphql({ schema, source: `
      query {
        test {
          resolverGotDefinitionInstance
        }
      }
    `});
    expect(result).toEqual({ data : {
      test: {
        resolverGotDefinitionInstance: true
      }
    }});
  });
});
