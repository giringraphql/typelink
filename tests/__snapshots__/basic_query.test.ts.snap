// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic queries and schema generation generates schema as expected 1`] = `
"type Query {
  test: Test
  erroneousTest: Test
  testPassingSource: Test
}

type Test {
  \\"\\"\\"description1\\"\\"\\"
  resolverGotDefinitionInstance: Boolean!
  greeting(greeting: String, name: String!): String!
  bigGreeting(greeting: String, name: String!): String!
  fieldWithDefaultResolver: String
}
"
`;
